### PR TITLE
OOP school library: refactor your code : module 4

### PR DESCRIPTION
**Pull request by yusufyusie**
# Project requirements

At this point your codebase is pretty big. The `main.rb` file may be especially large at this point. Let's try to clean things up!

- Make sure YAGNI, KISS, and DRY rules and Ruby style guide are respected.
- Follow SOLID principles.
- Pay special attention to the "Single-responsibility Principle".
    - Make sure that the role of `main.rb` is only to start your code and ask the user for the initial choice of options (or to quit).
    - Extract some pieces of logic into separate class(es).
    
#  In this milestone we've achieved the followings:
- with the help of my coding Partner I was able to fix some issues with the approach I took to writing my program.
- we followed SOLID principles and ensured that we write our code based on them.
- we applied DRY, YAGNI, and KISS Principles to improve our code structure.

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -1,39 +1,37 @@
...
-  action&.call
+      if (1..7).include?(choice)
+        handle_menu_option(choice, @books, @people, @rentals)
+        break if choice == 7
+      else
+        puts 'Invalid choice. Please select a valid option (1-7).'
+      end
+    end
+  end
 end
```
Comment (Qoosim) [inline main.rb]: - [ ] Great job refactoring your code in line with `Single-Responsibility Principle`. However, the main role of main.rb is only to start your code and ask the user for the initial choice of options (or to quit). Anything initializing should be done inside your `app.rb` or whatever file name you used. Kindly work more on the separation of concerns.
Your `main.rb` can then look something similar to this :point_down: 

```
# *the refactor contents of main.rb file
require "./app.rb"

def main
  app = App.new()

  response = nil

  puts "Welcome to School Library App!\n\n"
  while response != "7"
    puts "Please choose an option by entering a number:"
    puts "1 - List all books"
    puts "2 - List all people"
    puts "3 - Create a person"
    puts "4 - Create a book"
    puts "5 - Create a rental"
    puts "6 - List all rentals for a given person id"
    puts "7 - Exit"
    response = gets.chomp

    case response
    when "1"
      app.list_books()
    when "2"
      app.list_people()
    when "3"
      app.create_person()
    when "4"
      app.create_book()
    when "5"
      app.create_rental()
    when "6"
      app.list_rentals_for_person_id()
    when "7"
      puts "Thank you for using this app!"
    end

    puts "\n"
  end
end

main()

```

**Review by Qoosim — CHANGES_REQUESTED** [review]
#  *Status: Required Changes* :recycle:

Hi @team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
-  Implemented some project requirements. :heavy_check_mark: 
-  No linter errors. :heavy_check_mark: 
-  Good README file. :heavy_check_mark: 
-  Correct Gitflow. :heavy_check_mark: 




## Required Changes ♻️
_Check the comments under the review._

- [ ] I would have loved to highlight this section of your `app.rb` but since no refactor was done to that file, so it is not included among the files appeared where I can make use of inline commet to pass my feedback across to you. I therefore decided to make my comment here. But be aware that the comment I am making is meant for the `app.rb` file. Kindly apply whatever I say here to that file.

your-app


- [ ] First and foremost, Your `app.rb` file handles the responsibility of initializing your books, people, rental objects and creating methods that would call them outside the file to display them. All requests for input and input validation should've been placed in the classes/methods whose responsibilities are to request or validate input. None of those should be placed here for proper SOLID principle. You can create a class in a separate file and put all the methods that request for input and input validation. Then require them inside your `app.rb` and make use of them inside it. It should look something similar to this. :point_down: 

```
   require './book'
   require './rental'
   require './create'
   require './list'

   class App
     attr_reader :options

     def initialize(books_list = [], people_list = [])
         @books_list = books_list
         @people_list = people_list
         @create = Create.new
         @list = List.new
     end

     def list_books
       @list.books(@books_list)
        gets.chomp
     end

    def list_people
      @list.people(@people_list)
      gets.chomp
    end

    def create_person
      @people_list << @create.person
      gets.chomp
    end

    def create_book
      @books_list << @create.book
      gets.chomp
    end

    def create_rental
       @create.rental(@books_list, @people_list)
       gets.chomp
    end

    def list_rentals
       @list.rentals(@people_list)
       gets.chomp
    end
 end


  OR this :point_down: 

  
   require "./book.rb"
   require "./student.rb"
   require "./teacher.rb"
   require "./rental.rb"

   class App
      def initialize
        @books = []
        @people = []
        @rentals = []
     end
  
  # *Good example of Single responsability rule being followed
     def list_books
       @books.each do |book|
           puts book
       end
     end
  
    def list_people
       @people.each do |person|
           puts person
       end
    end
  end 

```

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: main.rb (hunk 1 above)
Comment (dvcorreia1984) [reply main.rb]: @Qoosim Thank you for you detailed feedback. We will work on this!

File: create_rental.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  person = select_person(people)
+
+  if book && person
+    date = rental_date
+    create_and_add_rental(rentals, date, book, person)
+    puts 'Rental created successfully'
+  else
+    puts 'Invalid input, please try again'
+  end
+end
```
Comment (V-Blaze) [inline create_rental.rb]: - You have done great so far :+1: But there are a few errors you still need to fix before moving to the next milestone. :+1: :smile: 
- [ ] Looking at the video below, you can see that the App crashes whenever I try to create a rental, this is not meant to be so. I suggest you look into this and fix this. Do well to thoroughly test the app before requesting another review. :+1: :100: :+1: 

- - VIDEO REFERENCE:

https://github.com/dvcorreia1984/oop-school-library/assets/107632780/3a9f7a62-f0da-4403-ad43-aa52e24ef553

**Review by V-Blaze — CHANGES_REQUESTED** [review]
## Hi @yusufyusie, @dvcorreia1984  

# Almost there 🟡 

giphy


Great work so far, You have come a long way, keep moving forward👏 👍  
You've done well implementing the project requirements, and some changes from the previous reviewers, but there are still some issues I think you need to improve on. 😄 👋 


 ## Highlights
- Code Follows SOLID principle✔
- Good communication with the reviewer :heavy_check_mark: 
- Correct git flow :heavy_check_mark: 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by danmainah — APPROVED** [review]
Hi @yusufyusie and @dvcorreia1984 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 

## Highlights
✔️  Implement the changes requested to fix the error when running the app.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**